benchexec/tools/sv_benchmarks_util.py

```diff
@@ -40,3 +44,196 @@ def get_data_model_from_task(task, param_dict):
                     f"Unsupported data_model '{data_model}' defined for task '{task}'"
                 )
     return None
+
+
+def _partition_input_files(input_files, task_options):
+    """
+    This function partitions the input files into witness and other files.
+    The distinction is based on the file name of the witness file, which
+    is identified by the option "witness" in the task options.
+
+    @param input_files: List of input files
+    @param task_options: Dictionary of task options
+    @return: Tuple of witness files and other files
+    """
+    witness_files = []
+    other_files = []
+    for file in input_files:
+        if Path(file).name == task_options.get("witness"):
+            witness_files.append(file)
+        else:
+            other_files.append(file)
+    return witness_files, other_files
+
+
+def get_witness(task):
+    """
+    This function returns the unique witness file from the task.
+    It raises an exception if there are multiple witness files.
+    The witness file is identified by the option "witness" in the task options.
+
+    @param task: An instance of a task
+    @return: Unique witness file
+    """
+    witness_files, _ = _partition_input_files(
+        task.input_files_or_identifier, task.options
+    )
+    if len(witness_files) != 1:
+        logging.warning(
+            "Tool uses benchexec default validation task processing. "
+            "It does not support tasks with more than one witness"
+        )
+        raise UnsupportedFeatureException(
+            "Tool only supports tasks with exactly one witness file"
```

Comment:
First, logging *and* throwing is an anti-pattern, one needs to do either of them, not both. If one does both, the two parts of the user message will get split and might be logged with several lines in between, or one line might not be logged at all. Furthermore, suppose we implement a wrapper that takes care of this situation and handles the exception, then we would not want any message to be shown to the user. But the wrapper could not avoid the log message.

Second, there is no "benchexec default validation task processing". This wording will lead people to think as if BenchExec somehow defines and implements how to handle validation tasks, but this is not true. BenchExec itself does not care or know or enforce anything about validation tasks. It is just the tool-info modules (most of which happen to be bundled with BenchExec) that do this.

Third, according to your [description](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1572) of these tasks, the `witness` option needs to be a single string with a path. So it is not even the tool-info module that does not support more than one witness, but your format extension.

Fourth, what happens here is not really about "one witness" or "more than one witness". As mentioned above, we only have one witness in the options. What has happened if we enter this case is that we encountered a task definition where there is one witness that occurs more than once in the input files, at least with the same base name:
```yaml
input_files: ["foo/witness.yml", "bar/witness.yml", "program.c"]
options:
  witness: "witness.yml"
```
Now, we could in principle allow this (just append each of the witness files to the command line), but I do not really see a reason why we would want this. It seems this is more likely an error in the task definition, and it is nicer for the user if we detect this.

Fifth, the current error message about "more than one witness" is also produced when the witness is missing from the input files. This needs to be taken into account when writing the message and in the documentation (where it is currently also wrong).

I think we should probably have a message that says something like the witness in the task definition did not occur exactly once in the input files of the task definition.

Most of these points also apply below.

Reply:
Thanks a lot for the amazing explanation!

To address the concerns I removed the extra logging and adjusted the error message into one similar to your suggestion in commits 833b33a9ebafe3b337c47000f28704ce8e7b0724 and 21021b244a2e6bc96408566994fb7fb94c0ded95

Reply:
@danieldietsch Which case specifically are you aiming at?

I would like to again note (cf. item 4 of my long comment above) that the original error message about "more than one witness" did not describe the situation correctly at all and thus lead the discussion into a wrong direction based on wrong assumptions. The case handled with this error message did *not* occur if more than one witness was present in the task definition, instead it occurred if someone created a task definition that lists the single witness twice in the input files (cf. my example above). This simply does not make sense, we do not want to support it, and it is the responsibility of the user to fix their task definition. The current error message correctly describes that.

For your current comment, I am sorry but I don't really understand all of it. What I can say is:

BenchExec has no responsibility here at all. Tool-info modules for validators that want to compete in the SV-COMP validation track need to handle task definitions with a `witness` key in the options, this is the responsibility of the tool-info module. In order to reduce the implementation and maintenance effort, we provide utility methods that make it easy to implement this, and tool-info modules can and should make use of them. And in order to push this forward, @marian-lingsch has volunteered to adapt all existing tool-info modules. Thus, after this PR the developers of the tool-info module do not need to do anything unless they change how their tool needs to be called, so I do not see any point where some responsibility is left. Furthermore, there is also no way to support a superset or a subset of the features currently being handled: The format extension proposed by @marian-lingsch supports task definitions with exactly one witness, a tool-info module can either support this format extension or not, but not partially. And tool-info modules also cannot support more than one witness file in tasks with @marian-lingsch's format extension, because that extension only uses a single string to define the witness.

I hope this helps. If you have further questions, please clarify.

Reply:
No, that helps. I guess my confusion comes from the naming and parts of the docs, e.g., https://github.com/sosy-lab/benchexec/blob/addcaa97e5181e029bc10d58edffc948a4019d96/benchexec/tools/sv_benchmarks_util.py#L70

You are right, this _only_ deals with witnesses specified with yaml `options/witness:` structure (how would one write that in one line?). 

The API and the current docs make it somewhat hard to distinguish these witnesses from the witnesses contained in the `input_files` part. The exception messages also do not distinguish these cases, e.g. 
```
         "Witness given in the task-definition did not occur exactly once "
            "in input files of the task definition"
```
The witnesses given via input_files are "witness given in the task-definition" as well. 

So, basically, distinguishing _the_ "option-witness" from the sequence of "input_files-witnesses" is the problem.

Reply:
I have attempted to clarify this in 25ca9b85.

Reply:
This looks very helpful and much better, thank you!